Post about types and programming languages (#6)
dafny-lang/blog: _posts/2023-06-30-types-and-programming-languages.markdown

---
layout: post
title:  "Types and Programming Languages"
date:   2023-06-30 18:00:00 +0100
author: Mikael Mayer

> usemathjax: true

Author: ```suggestion
```

---
<script type="text/javascript" async src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<link rel="stylesheet" href="/blog/assets/css/types-and-programming-languages.css">
<img class="clickable" id="img-intro" src="/blog/assets/images/type-and-programming-languages/introimage.png" alt="A type checker on the term If(True, 0, 1)" style="display:block;margin-left:auto;margin-right:auto;width:500px;max-width:95%;"/>
As recalled in the [last blog post]({{ site.baseurl }}{% post_url 2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny %}),
the last thing you want in a software business is that the program written by your developers goes wrong. And there are many ways programs can go wrong. One of these ways is of using a value in a way it was not intended to be used, and for that, static typing often solves entirely the problem.

This blog post takes a deep dive... take a deep breath... on how to write _terms of a programming language_, and both a _type-checker_ and an _evaluator_ on such terms, such that the following _soundness property_ holds:

> > 1. [Progress] If a type checker accepts a term, then the evaluator will _not get stuck_ on that term.

Author: No I made it up, so chances it existed before are greatly reduced and would be a coincidence. From the book from which I drew inspiration, quoted later, it's less user-friendly.

        IsZero(OneStepEvaluate(e))
  
    case Add(a, b) =>
      if !IsFinalValue(a) then
        Add(OneStepEvaluate(a), b)
      else if !IsFinalValue(b) then
        Add(a, OneStepEvaluate(b))
      else if a.Zero? then
        b
      else if b.Zero? then
        a
      else if a.Succ? then
        Succ(Add(a.e, b))
      else
        assert a.Pred?;
        Pred(Add(a.e, b))

    case Double(a) =>
      if IsFinalValue(a) then
        Add(a, a)
      else
        Double(OneStepEvaluate(a))

    case If(cond, thn, els) =>
      if IsFinalValue(cond) then
        assert GetType(cond) == Some(Bool);
        assert cond == True || cond == False;
        if cond == True then
          thn
        else
          assert cond == False;
          els
      else
        If(OneStepEvaluate(cond), thn, els)
  }
}
{% endhighlight %}

The interesting points to note on the function above, in a language like Dafny where every pattern must be exhaustive, are the following:

* Every call consist either of `OneStepEvaluate` on one sub-argument, or a transformation that reduces the size of the tree. So something is always happening here.
* All the cases are covered, Dafny does not complain!

>   * For example, when encountering the case `IsZero(e)`, if `e` is a final value, we know it's either `Pred` or `Succ`. It cannot be `True` or `False` due to type checking, so we can conclude.

Reviewer: ```suggestion
  * For example, when encountering the case `IsZero(e)`, if `e` is a final value, it must be either `Pred` or `Succ`. It cannot be `True` or `False` as it's well-typed and the previous pattern precludes `Zero`.
```

>   * Similarly, if the condition of an if term is a final value, because of type-checking, Dafny knows it's either True or False.

Reviewer: ```suggestion
  * Similarly, if the condition of an if term is a final value, because it's well-typed, Dafny knows it's either `True` or `False`.
```

> That concludes the _progress_ part on soundness checking: Whenever a term type-checks, it small step evaluation rule is never stuck until it reaches a final value.

Reviewer: ```suggestion
That concludes the _progress_ part of soundness checking: Whenever a term type-checks, there is always an applicable small step evaluation rule unless it's a final value.
```

#### The preservation check

> Soundness also contains another part, the preservation, as stated in the intro. It says that, if we evaluate a typed term, not only the evaluator is not going to get stuck, but the result will have the same type.

Reviewer: ```suggestion
Soundness has another aspect, preservation, as stated in the intro. It says that, when evaluating a well-typed term, the evaluator will not get stuck and the result will have the same type as the original term.
```

Dafny can also prove it for our language, out of the box. Well done, that means our language and evaluator make sense together!

{% highlight javascript %}
lemma OneStepEvaluateWellTyped(e: Term)
  requires WellTyped(e) && !IsFinalValue(e)
  ensures GetType(OneStepEvaluate(e)) == GetType(e)
{
  // Amazing: Dafny can prove this soundness property automatically
}
{% endhighlight %}

#### Conclusion

> All the code above powers this page, so that why I can guarantee you that you won't be able to find a term that the type checker accepts and that won't result in a final value. Of course, in a real programming language term, you might add some infinite loops here, but the soundness property above is not about termination, it's about constant progress, which you also want in embedded systems to ensure they never need reboot.

Reviewer: ```suggestion
All the code above powers this page, which is why I can guarantee you that you won't be able to find a term that the type checker accepts and that won't result in a final value. Of course, in a real programming language term, you might add some infinite loops, but the soundness property above is not about termination, it's about constant progress, which you also want in embedded systems to ensure they never need reboot.
```

> Now that you know what a type checker is and how to implement one in Dafny, perhaps you will feel much better prepared to model and experiment your new programming language, like recently the [Cedar team did](https://aws.amazon.com/about-aws/whats-new/2023/05/cedar-open-source-language-access-control/)?

Reviewer: ```suggestion
Now that you know what a type checker is and how to implement one in Dafny, perhaps you will feel much better prepared to model and experiment on your new programming language, like recently the [Cedar team did](https://github.com/cedar-policy/cedar-spec)?
```

This is the end of the blog post. I hope you enjoyed it so far, but if you are looking for some advanced concepts, feel free to continue reading! Beware, math ahead!

## Bonus: More advanced modeling

> Sometimes, modeling evaluator and type-checker as functions is not enough. One wants to model them as relations, and determine some properties about these relations, such as the order of evaluation does not matter for the final result.

Reviewer: ```suggestion
Sometimes, modelling the evaluator and the type-checker as functions is not enough. One wants to model them as relations, and determine some properties about these relations, such as the order of evaluation being irrelevant for the final result.
```

> In the rest of this blog post largely inspired by the book "Types and Programming Languages", Chapter 8, written by Benjamin Pierce, I will illustrate one element of the proof: the one that inductive and constructive versions of the set of Terms are equivalent. Having equivalence enables to conclude other results out of the scope of this blog post, including that the order of evaluation does not matter.

Reviewer: ```suggestion
In the rest of this blog post, largely inspired by the book "Types and Programming Languages", Chapter 8, written by Benjamin Pierce, I will illustrate one element of the proof: the one that inductive and constructive versions of the set of Terms are equivalent. Having equivalence enables obtaining other results out of the scope of this blog post, including that the order of evaluation does not matter.
```

> Once this proof trick is gained, it becomes possible to prove similar equivalences for different inductive and constructive definitions of:

Reviewer: ```suggestion
With the help of this trick, it becomes possible to prove similar equivalences for different inductive and constructive definitions of:
```

- The set of `(Expr, Expr)` of small-step evaluations
- The set of `(Expr, Type)` of type checking
but I leave these as an exercise for the interested reader.

In Types and Programming Languages, chapter 3.2, we discover that there are two other mathematical definitions of the "set of all terms".
The first one in definition 3.2.1 states that the set of _terms_ is the smallest set 𝒯 such that:

1. $$\{\texttt{true}, \texttt{false}, 0\} \subseteq 𝒯$$;
2. if $$t_1 \in 𝒯$$, then $$\{\texttt{succ}\;t_1, \texttt{pred}\;t_1, \texttt{is_zero}\;t_1\} \subseteq 𝒯$$;
3. if $$t_1 \in 𝒯$$, $$\;\; t_2 \in 𝒯$$ and $$t_3 \in 𝒯$$, then $$\texttt{if}\;t_1\;\texttt{then}\;t_2\;\texttt{else}\;t_3 \in 𝒯$$.

> Note that these terms omit `Double` and `Add` above. This means we cannot state that this set is the same as `set t: Term | true` as one would like to write. But let's continue.

Reviewer: ```suggestion
Note that these terms omit `Double` and `Add` above. This means we cannot state that this set is the same as `set t: Term | true` as one would like to write, but let's continue.
```

We can write the inductive definition above in Dafny too:

{% highlight javascript %}
ghost const AllTermsInductively: iset<Term>

ghost predicate InductionCriteria(terms: iset<Term>) {
  && iset{True, False, Zero} <= terms
  && (forall t1 <- terms
        :: iset{Succ(t1), Pred(t1), IsZero(t1)} <= terms)
  && (forall t1 <- terms, t2 <- terms, t3 <- terms
        :: If(t1, t2, t3) in terms)
}
lemma {:axiom} InductiveAxioms()
  ensures InductionCriteria(AllTermsInductively)
  ensures forall setOfTerms: iset<Term> | InductionCriteria(setOfTerms)
            :: AllTermsInductively <= setOfTerms
{% endhighlight %}

The second definition for the set of all terms in section 3.2.3 is done constructively. We first define a set $$S_i$$ for each natural number $$i$$, as follows

$$S_0 = ∅ $$;  

> $$\begin{aligned}S_{i+1} = && && \{\texttt{true}, \texttt{false}, 0\} \\ && \bigcup && \{\texttt{succ}\;t_1, \texttt{pred}\;t_1, \texttt{is_zero}\;t_1 \mid t_1 \in S_i \} \\ && \bigcup && \{\texttt{if}\;t_1\;\texttt{then}\;t_2\;\texttt{else}\;t_3\mid t_1 \in S_i,\; t_2 \in S_i,\; t_3 \in S_i\}\end{aligned}$$.

Reviewer: I'm not sure how to get rid of the dot.

Author: Let's leave it then. It was there in the original textbook.
